pandas/io/html.py

```diff
@@ -342,44 +345,149 @@ def _build_doc(self):
 
         Returns
         -------
-        obj : tree-like
+        obj : the DOM from which to parse the table element.
         """
         raise AbstractMethodError(self)
 
-    def _build_table(self, table):
-        header = self._parse_raw_thead(table)
-        body = self._parse_raw_tbody(table)
-        footer = self._parse_raw_tfoot(table)
+    def _build_table(self, table_html):
+        header, body, footer = self._parse_raw_thead_tbody_tfoot(table_html)
+        # the above "footer" actually produces a footer. The below "footer"
+        # rarely does. The below "footer" is the legacy behavior and so I'm
+        # leaving it for the time being.
+        footer = self._parse_raw_tfoot(table_html)
         return header, body, footer
 
-    def _parse_raw_thead(self, table):
-        thead = self._parse_thead(table)
+    def _parse_raw_thead_tbody_tfoot(self, table_html):
+        """Given a table, return parsed header, body, and foot.
+           Header and body are lists-of-lists. Top level list is a list of
+           rows. Each row is a list of parsed elements.
+
+           Logic: Use <thead>, <tbody>, <tfoot> elements to identify
+                  header, body, and footer, otherwise:
+                  - Put all rows into body
+                  - Move rows from top of body to header only if
+                    all elements inside row are <th>
+                  - Move rows from bottom of body to footer only if
+                    all elements inside row are <th>
+
+        Parameters
+        ----------
+        table_html : a single HTML table element.
+
+        Returns
+        -------
+        tuple of (header, body, footer)
+        header : list of rows, each of which is a list of parsed header
+        elements
+        body : list of rows, each of which is a list of parsed body elements
+        footer : list of rows, each of which is a list of parsed footer
+        elements
```

Comment:
Visually, this formatting looks weird.  Can you indent any subsequent lines in your description and avoid lone words on a line.  For example:

~~~python
header : list of rows, each of which is
         a list of header elements
~~~